src/extblkdev/ExtBlkDevInterface.h

```diff
@@ -0,0 +1,178 @@
+    //! name of device to be examined. It may be an logical volume that needs detection of underlying devices
+    std::string devname;
+    inline bool operator!=(const ExtBlkDevProfile& profile) {
+      return devname!=profile.devname;
+    }
+  };
+  inline std::ostream& operator<<(std::ostream& out, const ExtBlkDevProfile& profile) {
+    out << "{" << profile.devname  << "}";
+    return out;
+  }
+
+  class ExtBlkDevInterface {
+  public:
+    virtual ~ExtBlkDevInterface() {}
+
+    /**
+     * Initialize the instance according to the content of
+     * **profile**. The **ss** stream is set with debug messages or
+     * error messages, the content of which depend on the
+     * implementation.
+     *
+     * Return 0 on success or a negative errno on error. When
+     * returning on error, the implementation is expected to
+     * provide a human readable explanation in **ss**.
+     *
+     * @param [in] profile argument set used to detect device
+     * @param [out] ss contains informative messages when an error occurs
+     * @return 0 on success or a negative errno on error.
+     */
+    virtual int init(ExtBlkDevProfile &profile, std::ostream *ss) = 0;
+
+    /**
+     * Return the profile that was used to initialize the instance
+     * with the **init** method.
+     *
+     * @return the profile in use by the instance
+     */
+    virtual const ExtBlkDevProfile &get_profile() const = 0;
+
+    /**
+     * Return the name of the underlying device detected by **init** method
+     *
+     * @return the name of the underlying device
+     */
+    virtual const std::string& get_devname() const = 0;
+
+    /**
+     * Provide status of underlying physical storage after compression.
+     * The values denote effectively total and available capacity for 
+     * incompressible data.
+     * The **ss** stream is set with debug messages or
+     * error messages, the content of which depend on the
+     * implementation.
+     *
+     * Return 0 on success or a negative errno on error. When
+     * returning on error, the implementation is expected to
+     * provide a human readable explanation in **ss**.
+     *
+     * @param [out] total total capacity for incompressible data
+     * @param [out] avail available capacity for incompressible data
+     * @param [out] ss contains informative messages when an error occurs
+     * @return 0 on success or a negative errno on error.
+     */
+    virtual int get_thin_utilization(uint64_t *total, uint64_t *avail, std::ostream *ss) = 0;
+
+    /**
+     * Populate property map with meta data of device.
+     *
+     * @param [in] prefix prefix to be prepended to all map values by this method
+     * @param [in,out] pm property map of the device, to be extended by attributes detected by this plugin
+     * @return 0 on success or a negative errno on error.
+     */
+    virtual int collect_metadata(const std::string& prefix, std::map<std::string,std::string> *pm) = 0;
```

Comment:
You are right, my primary goal was not to change the existing interfaces and just make the device support extendable by using plugins. However the split into two functions still makes sense to me. 
collect_metadata is only called one per device open operation, and collects useful static data that should be presented as part of the static OSD state to an admin. The VDO presents only a small set of data, while our proprietary plugin for a hardware compression capable storage device provides a much richer set, including hardware version, serial numbers, firmware versions and device configuration. The only mandatory entry in this meta data should be the total physical storage capacity, but the VDO device prefixes that with VDO_, which makes it non-trivial to standardize and extract. But we probably should standardize that, but it may break existing code that specifically queries the VDO's capacity.
get_thin_utilization on the other hand is called constantly, about once per second. It provides only one dynamic data item, which is the currently available physical capacity of the device. The other value passed is the static total physical capacity, which is passed as a convenience, as extracting the value from the meta data map and keeping it around in the KernelDevice and FileStore code explicitly is a pain, especially with the plugin-specific prefixes.
I would be preferring to keep two functions, but maybe renaming them according to their use, e.g. get_static_metadata (fillin-in free form string map) and get_stat (filling-in a structure or class with the current state of the device + convenience data, extendable in the future).